tests/std/tests/VSO_1062649_overaligned_function/test.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+
+#ifndef __cpp_aligned_new
+#pragma error("This test is only valid when C++17 over-aligned allication is supported")
+#endif
+
+#pragma warning(disable : 4324) // "was padded due to alignment specifier", thanks, Captain
+
+struct alignas(16) overaligned_t {
+    char non_empty;
+
+    void operator()(const void* storage, std::size_t storage_size) const {
+        const unsigned char* storage_bytes = reinterpret_cast<const unsigned char*>(storage);
+        const unsigned char* this_bytes    = reinterpret_cast<const unsigned char*>(this);
+
+        // loop instead of range comparison to avoid UB when not in range
+        for (std::size_t i = 0; i < storage_size; i++) {
+            assert(storage_bytes + i != this_bytes);
+        }
+        assert(static_cast<std::size_t>(reinterpret_cast<std::uintptr_t>(this) % alignof(overaligned_t)) == 0);
+    }
+};
+
+static_assert(alignof(overaligned_t) == 16, "overaligned_t is not aligned as expected");

```

Comment:
This feels like it belongs in a compiler test instead of a library test to me, but I could live with it if you feel strongly otherwise. Could we define a symbolic name for the magic number 16 and use it here and in the definition of `overaligned_t` if we do keep this?

Reply:
I think it is an `ABI` contract that could change with future CPUs platforms.
So I thought we need a reminder to revisit this test in case of that.
I can remove it if you think otherwise.

Reply:
Ok, will remove after change accepted below

Reply:
> I think it is an ABI contract that could change with future CPUs platforms.

We declared the type as `struct alignas(16) overaligned_t`; consequently either its alignment is 16 ([[dcl.align]/4](http://eel.is/c++draft/dcl.align#4)) or - if its natural alignment would have been greater than 16 - the program is ill-formed per [[dcl.align]/5](http://eel.is/c++draft/dcl.align#5). This `static_assert` could only fire if the compiler doesn't properly implement those requirements.